### PR TITLE
Variable mutability for closures

### PR DESCRIPTION
**Pull request by NatashaWso2**
$subject

If a function-body-block refers to a block-scope variable defined outside of the function-body-block, the closure will capture a reference to that variable; the captured reference will refer to the same storage as the original reference not a copy.  This PR will pass the same storage as the original reference to the closure instead of a **copy of the original reference**.

Fixes https://github.com/ballerina-platform/ballerina-lang/issues/13849

## Automation tests
 - Unit tests 
   > Added

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 1]
```diff
@@ -1941,26 +2131,24 @@ public void visit(BLangFieldBasedAccess fieldAccessExpr) {
...
             if (fieldAccessExpr.symbol != null && fieldAccessExpr.symbol.type.tag == TypeTags.INVOKABLE
                     && ((fieldAccessExpr.symbol.flags & Flags.ATTACHED) == Flags.ATTACHED)) {
-                targetVarRef = new BLangStructFunctionVarRef(((BLangVariableReference) fieldAccessExpr.expr),
-                        (BVarSymbol) fieldAccessExpr.symbol);
+                targetVarRef = new BLangStructFunctionVarRef(fieldAccessExpr.expr, (BVarSymbol) fieldAccessExpr.symbol);
             } else {
                 targetVarRef = new BLangStructFieldAccessExpr(fieldAccessExpr.pos, fieldAccessExpr.expr, stringLit,
-                        (BVarSymbol) fieldAccessExpr.symbol, true);
+                        (BVarSymbol) fieldAccessExpr.symbol,
+                        true);
```
Comment (pranavan15) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]: Why we added a new line for the above? Shall we join this with the previous line as it was there earlier?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 2]
```diff
@@ -1914,6 +2067,43 @@ public void visit(BLangSimpleVarRef varRefExpr) {
...
+            // x = 1 ==> $innerMap$1["x"] = <any> 1
+            result = rewriteExpr(accessExpr);
+            return true;
+        }
+        // int z = x + 1 ==> int z = <int>$innerMap$1["x"] + 1;
+        result = rewriteExpr(addConversionExprIfRequired(accessExpr, varRefExpr.type));
+        return true;
+    }
+
+    private boolean isBasicType(BType bType) {
```
Comment (pranavan15) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]: Is this method used to check whether a type is a simple basic type? If that is the case, shall we use a simple `<=` boolean operator to check whether it is a simple basic type? Also, if this check is for simple basic type shall we modify the method name as `isSimpleBasicType`?

i.e: `bType.tag <= TypeTags.BOOLEAN_TAG`

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 3]
```diff
@@ -628,16 +689,59 @@ public void visit(BLangErrorVariable varNode) {
...
         block.stmts = rewriteStmt(block.stmts, blockEnv);
+
+        // Pop from stack
+        ((BLangFunction) env.enclInvokable).blockSymbolsInUpperLevels.pop();
         result = block;
     }
 
+    private void defineMapInBlockNode(BLangBlockStmt blockStmt) {
+        BLangRecordLiteral emptyRecord = ASTBuilderUtil.createEmptyRecordLiteral(blockStmt.pos,
+                symTable.mapType);
```
Comment (pranavan15) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]: Shall we join this line with the above?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 4]
```diff
@@ -550,22 +553,80 @@ public void visit(BLangFunction funcNode) {
...
-                .filter(symbol -> !isFunctionArgument(symbol, funcNode.symbol.params))
-                .forEach(symbol -> {
-                    symbol.closure = true;
-                    dupFuncSymbol.params.add(0, symbol);
-                    dupFuncType.paramTypes.add(0, symbol.type);
-                });
+        // Create a map symbol for the function statement
+        BUnionType constraintTypeOfMap = new BUnionType(null,
+                new LinkedHashSet<BType>() {{ add(symTable.anyType); add(symTable.errorType); }}, true);
+        BVarSymbol mapSymbol = new BVarSymbol(0, names.fromString("$mapFunc$" + funClosureMapCount),
```
Comment (pranavan15) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]: Shall we use string constants for the string literals? 
eg: "$mapFunc$", "$paramsMap$"

**Comment by NatashaWso2** [comment]
Closing this PR since this needs some significant improvements. Will open another PR after the changes are done.

**Pull request closed without merging**